### DISCUSSION TITLE
Thoughts on prefixing tactic applications with 'all:'

### DISCUSSION THREAD

**Discussion by magmkri** [princeton-vl/CoqGym#51] [General]
Hi!

As I understand it tactics are by default only applied to the first subgoal in a given proof state. However, this can be changed by superseding the tactic with the `all:` keyword. But because of the `timeout` included when sending tactics via `eval_env.py` the tactic application fails.
https://github.com/princeton-vl/CoqGym/blob/cde1f3f2cd0fa208ea44085e26cae269eec2bc1b/eval_env.py#L76
It turns out `all:` has to come before `timout`. A simple, ad-hoc, fix is:  
```
if "all:" in command[:-1]:
	command = "all: timeout %d (%s)." % (time_left, command[:-1].replace("all: ", ""))
else:
	command = "timeout %d (%s)." % (time_left, command[:-1])
```

Drawing the proof tree for a (simple) agent's proof, this seems shorten the proof search somewhat:
![no 'all:'](https://github.com/MaganMK/CoqGym/blob/gnn/sandbox/pngs/ffn/trans_eeq.png)
![with 'all:'](https://github.com/MaganMK/CoqGym/blob/gnn/sandbox/pngs/ffn/trans_eeq2.png)

Any thoughts on adding support for `all:` in `ProofEnv`? :)

As a related sidenote to this:
It also seems that checking only the first (sub)goal signature can make agent's struggle to prove certain theorems:
https://github.com/princeton-vl/CoqGym/blob/cde1f3f2cd0fa208ea44085e26cae269eec2bc1b/ASTactic/agent.py#L304
Certain proof trees can become simpler by changing this to checking the concatenation of all subgoal signatures:
![checking first signature](https://github.com/MaganMK/CoqGym/blob/gnn/sandbox/pngs/ffn/eeq_refl.png)
![checking all signatures](https://github.com/MaganMK/CoqGym/blob/gnn/sandbox/pngs/ffn/eeq_refl2.png)
But again: in these kinds of proof states using `all:` fixes this implicitally, and the result is an even shorter proof tree:
![using 'all:'](https://github.com/MaganMK/CoqGym/blob/gnn/sandbox/pngs/ffn/eeq_refl3.png)

Hope this makes sense :). Thanks in advance!

**Comment by yangky11** [comment]
Hi Magus,

That's an interesting point. But I cannot see the figures. Could you fix them?

**Reply by magmkri** [reply]
Hi,

Yes, of course! I think they should work now. Just in case, I'm attaching the last three ('checking first signature', 'checking all signatures', and 'using :all') here, as they illustrate the point. This is of course not a major thing at all – it was more something I was curious about, as it seems to be able to impact the search tree somewhat :). 

Btw, if it seems strange why the yellow path in the first image does not explore the use of 'eauto', I belive it is because that would result in a (first subgoal) signature identical to the one obtained after applying 'constructor' to the root (hence why checking for a concatenation of the subgoal signatures generates a simpler proof tree in the second image: each state is "more strictly unique"). (The colors along the edges are not important. They just signal whenever the depth first search backtracks, and explores a new path). 

![eeq_refl](https://user-images.githubusercontent.com/32375836/114463779-2d043a80-9be5-11eb-95ca-b73761db433c.png)
![eeq_refl2](https://user-images.githubusercontent.com/32375836/114463792-32618500-9be5-11eb-94c3-8dea761cb90b.png)
![eeq_refl3](https://user-images.githubusercontent.com/32375836/114463799-355c7580-9be5-11eb-89f2-ecae402ee9a2.png)

**Reply by yangky11** [reply]
Thanks! How does the agent decide whether or not to use `all:`? Does it use `all:` for all tactics?

**Reply by magmkri** [reply]
I simply allowed the agent to use `all:` for all tactics, because I was mostly curious to see what impact it had out-of-the-box. But even though this might help an agent in the proof search, I decided not to include it in the agents I'm testing now. It didn't seem that interesting to help an agent using  a "trick" likes this, when the goal is see the impact of different embedding strategies, training paradigms, etc. 
 
However, perhaps it would be more interesting to include `all:` if it was something the agent had to make a decision about. For example having a model dedicated to predicting the usefullness of `all:`, or if using a context free grammar (as ASTactic does) perhaps somehow include it in the grammar.  Of course, the agent would have to embedd not just a the first subgoal in a given state, but all goals in the state when deciding what to do with `all:`.

**Reply by yangky11** [reply]
I see, makes sense. I think for now people can just use the hacky way to support `all:` as you did earlier. And we'll see what's the best way of incorporating it if it turns out to be useful.

**Reply by magmkri** [reply]
Absolutely 👍🏽 

Thanks!
